src/hooks/users.ts

```diff
@@ -1,63 +1,159 @@
-import { useFirestoreConnect } from "hooks/useFirestoreConnect";
+import { useSelector } from "./useSelector";
+import { useUserLastSeenThreshold } from "./useUserLastSeenThreshold";
+import { useConnectCurrentVenueNG } from "./useConnectCurrentVenueNG";
+import { useVenueId } from "./useVenueId";
+import { useFirestoreConnect, isLoaded } from "./useFirestoreConnect";
 
-const useConnectVenueUsers = () => {
+export const useConnectWorldUsers = () => {
   const venueId = useVenueId();
 
-  // @debt refactor this + related code so as not to rely on using a shadowed 'storeAs' key
-  //   this should be something like `storeAs: "currentVenueUsers"` or similar
-  useFirestoreConnect(
+  const { isCurrentVenueLoaded, currentVenue } = useConnectCurrentVenueNG(
     venueId
-      ? {
-          collection: "users",
-          where: ["enteredVenueIds", "array-contains", venueId],
-          storeAs: "users" as ValidStoreAsKeys, // @debt super hacky, but we're consciously subverting our helper protections
-        }
-      : undefined
+  );
+
+  useFirestoreConnect(() => {
+    if (!isCurrentVenueLoaded || !currentVenue || !venueId) return [];
+
+    const relatedLocationIds = [currentVenue.id];
+
+    if (currentVenue.parentId) {
+      relatedLocationIds.push(currentVenue.parentId);
+    }
+
+    return [
+      {
+        collection: "users",
+        where: ["enteredVenueIds", "array-contains-any", relatedLocationIds],
+        storeAs: "worldUsers",
+      },
+    ];
+  });
+};
+
+export const useWorldUsers = (): {
+  worldUsers: readonly WithId<User>[];
+  isWorldUsersLoaded: boolean;
+} => {
+  useConnectWorldUsers();
+
+  const selectedWorldUsers = useSelector(worldUsersSelector);
+
+  return useMemo(
+    () => ({
+      worldUsers: selectedWorldUsers ?? [],
+      isWorldUsersLoaded: isLoaded(selectedWorldUsers),
+    }),
+    [selectedWorldUsers]
   );
 };
 
-export const useVenueUsers = (): readonly WithId<User>[] => {
-  useConnectVenueUsers();
+export const useWorldUsersById = () => {
+  useConnectWorldUsers();
+
+  const worldUsersById = useSelector(worldUsersByIdSelector);
 
-  return useSelector(usersSelector) ?? [];
+  return useMemo(
+    () => ({
+      worldUsersById: worldUsersById ?? {},
+      isWorldUsersLoaded: isLoaded(worldUsersById),
+    }),
+    [worldUsersById]
+  );
 };
 
-export const usePartygoers = (): readonly WithId<User>[] => {
+export const useRecentWorldsUsers = (): {
+  recentWorldUsers: readonly WithId<User>[];
+  isRecentWorldUsersLoaded: boolean;
+} => {
   const lastSeenThreshold = useUserLastSeenThreshold();
 
-  const venueUsers = useVenueUsers();
+  const { worldUsers, isWorldUsersLoaded } = useWorldUsers();
 
   return useMemo(
-    () => venueUsers.filter((user) => user.lastSeenAt > lastSeenThreshold),
-    [venueUsers, lastSeenThreshold]
+    () => ({
+      recentWorldUsers: worldUsers.filter(
+        (user) => user.lastSeenAt > lastSeenThreshold
+      ),
```

Comment:
As an interim 'hacky fix' to work around the fact that some data is in seconds, and some data is in milliseconds, we could have this function 'normalise' them to milliseconds first, then do the comparison.

- eg. if `user.lastSeenAt` is less than or equal to `MAX_SECONDS_VALUE`, then multiply it by 1000 before checking if it is greater than `lastSeenThreshold`

Probably would mark this with `@debt` as something for us to come back to/fix properly once we'd fixed/aligned the data in firestore.

Reply:
Since this is related(https://github.com/sparkletown/sparkle/pull/1047#discussion_r563651191), I could just set every place to use milliseconds. It will not break anything(only if there are events happening at the very time of merging this to production). This seems to be the most efficient way to solve this. Is there any danger in going this way?

Reply:
When you say ‘just set every place to use milliseconds’, what do you mean?

There’s still going to be the case of handling data that already exists in firestore that isn’t stored in milliseconds. So unless you’re planning on backfilling all of that data as well, then this is probably still required.

The safest way to handle these situations is to make the code change that can handle both, then backfill the data, then remove the triage/workaround once we know the data is all clean/etc. We may not need that level of safety with the scale/etc of our app at the moment though.

Reply:
I mean `lastSeenIn` and `lastSeenAt`. What is the case you are referring to?

Reply:
My desire is for both to be in milliseconds eventually, as that is javascripts unit of time.

Reply:
> My desire is for both to be in milliseconds eventually, as that is javascripts unit of time.

Yes, I do totally agree with as it will also solve `old users visibillity` issue. I'm asking if there can be any drawbacks of just replacing every `lastSeenIn` and `lastSeenAt` usage with milliseconds without any data backfilling

Reply:
If you don't backfill, and you don't implement something like I mentioned in https://github.com/sparkletown/sparkle/pull/1047#discussion_r563655929 that can proactively handle both cases, then there will most likely be issues.

But if you do implement something that can handle both seconds and milliseconds wherever we 'read' that data, and then change everywhere that 'writes' to that data to use milliseconds, that will be able to handle the legacy data without us needing to backfill + also ensure our future data moves towards our desired state.